gadal/gadal_types/data_types.py

```diff
@@ -41,144 +47,297 @@ class CostArray(BaseModel):
 
+
 class Incidence(BaseModel):
-    from_equipment: List[str]    
-    to_equipment:    List[str]    
+    from_equipment: List[str]
+    to_equipment: List[str]
     equipment_type: Optional[List[str]]
 
+
 class AdmittanceSparse(Incidence):
     admittance_list: List[Complex]
-    units:  str = 'S'
+    units: str = "S"
+
 
 class AdmittanceMatrix(BaseModel):
     admittance_matrix: List[List[Complex]]
     ids: List[str]
-    units: str = 'S'
+    units: str = "S"
+
 
 class Injection(BaseModel):
-    current_real: CurrentsReal = {"values":[], "ids":[]}
-    current_imaginary: CurrentsImaginary = {"values":[], "ids":[]}
-    power_real: PowersReal = {"values":[], "ids":[]}
-    power_imaginary: PowersImaginary = {"values":[], "ids":[]}
-    impedance_real: ImpedanceReal = {"values":[], "ids":[]}
-    impedance_imaginary: ImpedanceImaginary = {"values":[], "ids":[]}
+    # Shouldn't these be equipment arrays?
+    current_real: CurrentsReal = {"values": [], "ids": [], "node_ids": []}
+    current_imaginary: CurrentsImaginary = {"values": [], "ids": [], "node_ids": []}
+    power_real: PowersReal = {"values": [], "ids": [], "node_ids": []}
+    power_imaginary: PowersImaginary = {"values": [], "ids": [], "node_ids": []}
+    impedance_real: ImpedanceReal = {"values": [], "ids": [], "node_ids": []}
+    impedance_imaginary: ImpedanceImaginary = {"values": [], "ids": [], "node_ids": []}
+
 
 Topology.update_forward_refs()
+
+
+class Command(BaseModel):
+    """JSON Configuration for external object commands.
+
+    obj_name -- name of the object.
+    obj_prop -- name of the property.
+    obj_val -- val of the property.
+    """
+
+    obj_name: str
+    obj_property: str
+    val: Union[int, float, str, List[int], List[float], List[str]]
+
+
+class CommandList(BaseModel):
+    """List[Command] with JSON parsing."""
+
+    __root__: List[Command]
+
+
+class ReactivePowerSetting(Enum):
+    """Reactive power setting, almost always VARAVAL_WATTS."""
+
+    VARAVAL_WATTS = "VARAVAL_WATTS"
+    VARMAX_VARS = "VARMAX_VARS"
+    VARMAX_WATTS = "VARMAX_WATTS"
+
+
+class InverterControlMode(Enum):
+    """Inverter control mode."""
+
+    voltvar = "VOLTVAR"
+    voltwatt = "VOLTWATT"
+    voltvar_voltwatt = "VV_VW"
+
+
+class VVControl(BaseModel):
+    """OpenDSS setting for volt-var control."""
+
+    deltaq_factor: float = 0.7
+    varchangetolerance: float = 0.025
+    voltagechangetolerance: float = 0.0001
+    vv_refreactivepower: ReactivePowerSetting = ReactivePowerSetting.VARAVAL_WATTS
+    voltage: List[float]  # p.u. in V
+    reactive_response: List[float]  # p.u. in VArs
+
+
+class VWControl(BaseModel):
+    """OpenDSS setting for volt-watt control."""
+
+    deltap_factor: float = 1.0
+    voltage: List[float]  # p.u. in V
+    power_response: List[float]  # p.u. in VArs
+
+
+class InverterControl(BaseModel):
+    """InverterControl with volt-var control and/or volt-watt control."""
+
+    pvsystem_list: Optional[List[str]] = None
```

Comment:
Where do you validate that if the if the PV pvsystem_list is a list it is of size 1?

Reply:
I currently do that in the feeder instead, since it could also be none to specify all PV systems.